stdlib/Logging/src/ConsoleLogger.jl

```diff
@@ -118,8 +118,17 @@ function handle_message(logger::ConsoleLogger, level::LogLevel, message, _module
     end
 
     # Generate a text representation of the message and all key value pairs,
-    # split into lines.
-    msglines = [(indent=0, msg=l) for l in split(chomp(convert(String, string(message))::String), '\n')]
+    # split into lines.  This is specialised to improve type inferesence,
+    # and reduce the risk of resulting method invalidations.
+    message = string(message)
+    msglines = if Base._isannotated(message)
+        message = Base.AnnotatedString(String(message), Base.annotations(message))
```

Comment:
Yes, seems right to me. I think you can also combine a few more code paths like this also
```suggestion
    message = string(message)
    messagestr = String(message)::String
    annots = Base._isannotated(message)
    messageattrs = annots ? Base.annotations(message) : ()
    msglines = if annots && !isempty(messageattrs)
        message = Base.AnnotatedString(messagestr, messageattrs)
```

Reply:
I can't help but feel this is taking DRY too far...

Reply:
I guess mainly it is ugly since I am taking advantage of the DRY-er code to have it use the non-annotated code path if it detects there are no relevant annotations on this (sub)string. I don't know if it is necessarily worth doing that, as it does force the code to be somewhat more linear here (longer) as each line cannot do multiple computations anymore (even just `annotations` and `String`).

Reply:
In the push I recently did, I added that check too by just doing `if Base._isannotated(message) && !isempty(Base.annotations(message))`. FWWIW, I think the main visual impediment with the current code isn't the fact it could be slightly DRY-er but the `Base.` prefixes. Without them, this is the current code:

```julia
message = string(message)
msglines = if _isannotated(message) && !isempty(annotations(message))
    message = AnnotatedString(String(message), annotations(message))
    @NamedTuple{indent::Int, msg::Union{SubString{AnnotatedString{String}}, SubString{String}}}[
        (indent=0, msg=l) for l in split(chomp(message), '\n')]
else
    [(indent=0, msg=l) for l in split(
         chomp(convert(String, message)::String), '\n')]
end
```